### PR TITLE
Deprecating `levenstein` in favor of `levensHtein`

### LINKED ISSUE #27325: Suggester string_distance option typo in code
The string_distance option for suggesters has 5 options. The default is ```demarau_levenshtein```. That's okay. One of the other options is the more classic ```levenstein```. That's without the `h`. I think the Levenshtein of Demarau-Levenshtein is the same dude who came up with the LevensHtein distance.

I thought maybe it was just a typo in the documentation, but I just tested it on 6.0.0-rc1.

screen shot 2017-11-08 at 8 03 23 pm

I'm guessing not many people use this option so nobody noticed.

### PR DESCRIPTION
**Pull request by olcbean**
Fixes #27325

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Thanks @olcbean but I'm afraid it's a bit more complicated than this. This is a breaking change, so we need to support both spellings, and emit a deprecation warning for the incorrect one, before the incorrect one can be removed in a subsequent major release.

**Comment by olcbean** [comment]
@DaveCTurner do you mean to deprecate it for `6.1`?
Just one more question : do you prefer to have 2 PR against `master` (deprecation and removal) or 1 against `6.1` and one against `master`

**Comment by rjernst** [comment]
@olcbean 2 PRs would be normal, do the deprecation first, then removal in master as a followup.

**Comment by DaveCTurner** [comment]
Yes, 2 PRs please, both against `master`, and we will backport the deprecation one to `6.x`: if the deprecation PR was only merged into `6.x` and then the removal PR stalled for some reason then we'd have a regression.

Anticipating future feedback (because I forgot last time I did something similar): there should be a test that the deprecation warning is emitted on the incorrect spelling.

**Comment by olcbean** [comment]
Thanks @DaveCTurner 
I just deprecated the `levenstein`

File: core/src/main/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorBuilder.java [hunk 1]
```diff
@@ -458,13 +463,16 @@ private static SuggestMode resolveSuggestMode(String suggestMode) {
...
         if ("internal".equals(distanceVal)) {
             return DirectSpellChecker.INTERNAL_LEVENSHTEIN;
         } else if ("damerau_levenshtein".equals(distanceVal) || "damerauLevenshtein".equals(distanceVal)) {
             return new LuceneLevenshteinDistance();
         } else if ("levenstein".equals(distanceVal)) {
+            DEPRECATION_LOGGER.deprecated("Deprecated distance [levenstein] used, replaced by [levenshtein]");
+            return new LevensteinDistance();
+        } else if ("levenshtein".equals(distanceVal)) {
             return new LevensteinDistance();
             // TODO Jaro and Winkler are 2 people - so apply same naming logic
```
Comment (DaveCTurner) [inline core/src/main/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorBuilder.java]: It seems a shame to leave this TODO here. Would you care to fix that too?

**Comment by DaveCTurner** [comment]
Looking good. I added a request for some scope creep :)

@elasticmachine please test this.

File: core/src/main/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorBuilder.java (hunk 1 above)
Comment (jasontedor) [reply core/src/main/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorBuilder.java]: No scope creep please. :smile:

File: core/src/test/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorTests.java [hunk 2]
```diff
@@ -65,6 +67,11 @@ public void testEqualsAndHashcode() throws IOException {
         }
     }
 
+    public void testLevensteinDeprecation() {
+        assertThat(DirectCandidateGeneratorBuilder.resolveDistance("levenstein"), equalTo(new LevensteinDistance()));
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorTests.java]: I might be missing something but I do not see a test that `resolveDistance("levenshtein")` does the right thing and does not produce a deprecation warning?

**Review by jasontedor — COMMENTED** [review]
I left a comment.

File: core/src/test/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorTests.java (hunk 2 above)
Comment (olcbean) [reply core/src/test/java/org/elasticsearch/search/suggest/phrase/DirectCandidateGeneratorTests.java]: Good catch!

**Comment by olcbean** [comment]
@DaveCTurner I will clean the Jaro and Winkler TODO once this PR is merged (I want to reuse the test I introduced in this PR)

**Comment by jasontedor** [comment]
test this please

**Comment by jasontedor** [comment]
retest this please

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by jasontedor** [comment]
@DaveCTurner If this looks good to you, would you mind pulling this in? I'm happy to walk you through the process, I think this would be your first time pulling a community PR in?

**Comment by DaveCTurner** [comment]
LGTM, thanks @olcbean!

Backported to `6.x` as 19afeafcb453aedb9c4777a18fc91ab4a6bd5de8 but not to `6.1` as that branch is now feature-frozen.

**Pull request merged**